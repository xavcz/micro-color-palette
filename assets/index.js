// Native
const path = require('path');
const http = require('http');
const fs = require('fs');

// Ours
const types = require('../src/types');

// exports a simple asset server
module.exports = http.createServer((req, res) => {
  // get the image absolute local path & its type
  const file = path.join(__dirname, req.url);
  const fileType = types[path.extname(file).slice(1)];

  const stream = fs.createReadStream(file);

  // stream it...
  stream.on('open', function() {
    res.setHeader('Content-Type', fileType);
    stream.pipe(res);
  });

  // ...or fail!
  stream.on('error', function() {
    res.setHeader('Content-Type', 'text/plain');
    res.statusCode = 404;
    res.end('Not found');
  });
});
