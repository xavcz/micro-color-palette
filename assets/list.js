// Native
const fs = require('fs');
const path = require('path');

// Returns all images from this directory in a SYNC (blocking) way
module.exports = () => fs.readdirSync(__dirname).filter(file => path.extname(file) !== '.js');

// ðŸ˜¢ the code below (asynchronous) cannot be runned by Jest directly
// Returns a promise that resolve in all images from this directory
// (= read the directory and exclude the .js files)
// module.exports = new Promise(resolve => {
//   fs.readdir(__dirname, (err, files) => {
//     resolve(files.filter(file => path.extname(file) !== '.js'));
//   });
// });
