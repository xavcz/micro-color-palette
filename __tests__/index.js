const { describe, it, expect, beforeAll, afterAll } = global;
const micro = require('micro');
const listen = require('test-listen');
const fetch = require('node-fetch');

// services definitions
const services = [
  // micro color palette service to test
  micro(require('../src/index')),
  // assets service to serve test images
  require('../assets'),
];

// async function to run a server from a service
const runServer = async service => {
  const url = await listen(service);

  return { service, url };
};

// clean servers for the tests
let servers;

beforeAll(async () => {
  servers = await Promise.all(services.map(runServer));
});

afterAll(() => {
  servers.forEach(server => server.service.close());
});

describe('image types', () => {
  it('works on png', async () => {
    const [micro, assets] = servers;

    const res = await fetch(micro.url, {
      method: 'POST',
      body: JSON.stringify({
        src: assets.url + '/micro.png',
      }),
    });

    const palette = await res.json();

    expect(palette).toMatchSnapshot();
  });

  it('works on gif', async () => {
    const [micro, assets] = servers;

    const res = await fetch(micro.url, {
      method: 'POST',
      body: JSON.stringify({
        src: assets.url + '/parrot.gif',
        type: 'gif',
      }),
    });

    const palette = await res.json();

    expect(palette).toMatchSnapshot();
  });

  it('works on jpg', async () => {
    const [micro, assets] = servers;

    const res = await fetch(micro.url, {
      method: 'POST',
      body: JSON.stringify({
        src: assets.url + '/koopa.jpg',
        type: 'jpg',
      }),
    });

    const palette = await res.json();

    expect(palette).toMatchSnapshot();
  });

  it('works on svg', async () => {
    const [micro, assets] = servers;

    const res = await fetch(micro.url, {
      method: 'POST',
      body: JSON.stringify({
        src: assets.url + '/walle.svg',
        type: 'svg+xml',
      }),
    });

    const palette = await res.json();

    expect(palette).toMatchSnapshot();
  });
});
