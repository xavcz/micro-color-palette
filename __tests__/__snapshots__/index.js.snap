// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image types works on gif 1`] = `
Array [
  "#ccb9f3",
  "#fc948c",
  "#2a440c",
  "#8cfc8c",
  "#7c8c6c",
]
`;

exports[`image types works on jpg 1`] = `
Array [
  "#d0694b",
  "#101111",
  "#6a3b2c",
  "#569988",
  "#407f86",
]
`;

exports[`image types works on png 1`] = `
Array [
  "#3494fc",
  "#64acfc",
  "#0474fc",
  "#8cc4fc",
  "#3cb0fc",
]
`;

exports[`image types works on svg 1`] = `
Array [
  "#394848",
  "#a2b2b3",
  "#636767",
  "#394848",
  "#a2b2b3",
  "#ddeced",
  "#ddeced",
  "#a2b2b3",
  "#ddeced",
  "#0c1112",
  "#394848",
  "#ffffff",
  "#f49123",
  "#a2b2b3",
  "#ddeced",
  "#0c1112",
  "#394848",
  "#ffffff",
  "#394848",
  "#636767",
  "#0c1112",
  "#0c1112",
  "#c6690c",
  "#a2b2b3",
  "#636767",
  "#636767",
  "#424e4f",
  "#424e4f",
  "#394848",
  "#424e4f",
  "#424e4f",
  "#f49123",
  "#ddeced",
  "#ddeced",
  "#636767",
  "#f2791e",
  "#f2791e",
  "#f2791e",
  "#ddeced",
  "#a2b2b3",
  "#bedb45",
  "#636767",
  "#f76f6f",
  "#a2b2b3",
  "#f2791e",
  "#a2b2b3",
  "#636767",
  "#636767",
  "#0c1112",
  "#394848",
  "#a2b2b3",
  "#636767",
  "#394848",
  "#394848",
  "#a2b2b3",
  "#ddeced",
  "#394848",
  "#a2b2b3",
  "#a2b2b3",
  "#ddeced",
  "#394848",
  "#a2b2b3",
  "#a2b2b3",
  "#ddeced",
  "#394848",
  "#a2b2b3",
  "#a2b2b3",
  "#ddeced",
  "#394848",
  "#a2b2b3",
  "#f2791e",
  "#ddeced",
  "#636767",
  "#636767",
  "#636767",
  "#a2b2b3",
  "#636767",
  "#a2b2b3",
  "#636767",
  "#0c1112",
  "#636767",
]
`;
